stacks-signer/src/signer.rs

```diff
@@ -497,20 +513,59 @@ impl Signer {
     }
 
     /// Handle proposed blocks submitted by the miners to stackerdb
-    fn handle_proposed_blocks(&mut self, stacks_client: &StacksClient, blocks: &[NakamotoBlock]) {
-        for block in blocks {
-            // Store the block in our cache
-            self.signer_db
-                .insert_block(&BlockInfo::new(block.clone()))
-                .unwrap_or_else(|e| {
-                    error!("{self}: Failed to insert block in DB: {e:?}");
-                });
-            // Submit the block for validation
-            stacks_client
-                .submit_block_for_validation(block.clone())
-                .unwrap_or_else(|e| {
-                    warn!("{self}: Failed to submit block for validation: {e:?}");
-                });
+    fn handle_proposed_blocks(
+        &mut self,
+        stacks_client: &StacksClient,
+        proposals: &[BlockProposalSigners],
+    ) {
+        for proposal in proposals {
+            if proposal.reward_cycle != self.reward_cycle {
+                debug!(
+                    "{self}: Received proposal for block outside of my reward cycle, ignoring.";
+                    "proposal_reward_cycle" => proposal.reward_cycle,
+                    "proposal_burn_height" => proposal.burn_height,
+                );
+                continue;
+            }
+            let sig_hash = proposal.block.header.signer_signature_hash();
+            match self.signer_db.block_lookup(self.reward_cycle, &sig_hash) {
+                Ok(Some(block)) => {
+                    debug!(
+                        "{self}: Received proposal for block already known, ignoring new proposal.";
+                        "signer_sighash" => %sig_hash,
+                        "proposal_burn_height" => proposal.burn_height,
+                        "vote" => ?block.vote.as_ref().map(|v| {
+                            if v.rejected {
+                                "REJECT"
+                            } else {
+                                "ACCEPT"
+                            }
+                        }),
+                        "signed_over" => block.signed_over,
```

Comment:
Same as my previous nit: If we implemented `Display` for the BlockInfo type we could reuse the implementation here and avoid some duplication.